File: appupdate/src/main/java/com/azhon/appupdate/manager/DownloadManager.java
package com.azhon.appupdate.manager;

import android.content.Context;
import android.content.Intent;
import android.text.TextUtils;
import android.widget.Toast;

import com.azhon.appupdate.R;
import com.azhon.appupdate.activity.PermissionActivity;
import com.azhon.appupdate.config.UpdateConfiguration;
import com.azhon.appupdate.dialog.UpdateDialog;
import com.azhon.appupdate.service.DownloadService;
import com.azhon.appupdate.utils.ApkUtil;
import com.azhon.appupdate.utils.Constant;
import com.azhon.appupdate.utils.LogUtil;
import com.azhon.appupdate.utils.PermissionUtil;

/**
 * 项目名:    AppUpdate
 * 包名       com.azhon.appupdate.manager
 * 文件名:    DownloadManager
 * 创建时间:  2018/1/27 on 10:27
 * 描述:     TODO
 *
 * @author 阿钟
 */


public class DownloadManager {

    private static final String TAG = Constant.TAG + "DownloadManager";

    /**
     * 上下文
     */
    private static Context context;
    /**
     * 要更新apk的下载地址
     */
    private String apkUrl = "";
    /**
     * apk下载好的名字 .apk 结尾
     */
    private String apkName = "";
    /**
     * apk 下载存放的位置
     */
    private String downloadPath;
    /**
     * 是否提示用户 "当前已是最新版本"
     * <p>
     * {@link #download()}
     */
    private boolean showNewerToast = false;
    /**
     * 通知栏的图标 资源路径
     */
    private int smallIcon = -1;
    /**
     * 整个库的一些配置属性，可以从这里配置
     */
    private UpdateConfiguration configuration;
    /**
     * 要更新apk的versionCode
     */
    private int apkVersionCode = 1;
    /**
     * 显示给用户的版本号
     */
    private String apkVersionName = "";
    /**
     * 更新描述
     */
    private String apkDescription = "";
    /**
     * 安装包大小 单位 M
     */
    private String apkSize = "";
    /**
     * 兼容Android N 添加uri权限 authorities
     */
    private String authorities = "";

    private static DownloadManager manager;

    /**
     * 框架初始化
     *
     * @param context 上下文
     * @return {@link DownloadManager}
     */
    public static DownloadManager getInstance(Context context) {
        DownloadManager.context = context;
        if (manager == null) {
            synchronized (DownloadManager.class) {
                if (manager == null) {
                    manager = new DownloadManager();
                }
            }
        }
        return manager;
    }

    /**
     * 供此依赖库自己使用.
     *
     * @return {@link DownloadManager}
     * @hide
     */
    public static DownloadManager getInstance() {
        if (manager == null) {
            throw new RuntimeException("请先调用 getInstance(Context context) !");
        }
        return manager;
    }

    /**
     * 获取apk下载地址
     */
    public String getApkUrl() {
        return apkUrl;
    }

    /**
     * 设置apk下载地址
     */
    public DownloadManager setApkUrl(String apkUrl) {
        this.apkUrl = apkUrl;
        return this;
    }

    /**
     * 获取apk的VersionCode
     */
    public int getApkVersionCode() {
        return apkVersionCode;
    }

    /**
     * 设置apk的VersionCode
     */
    public DownloadManager setApkVersionCode(int apkVersionCode) {
        this.apkVersionCode = apkVersionCode;
        return this;
    }

    /**
     * 获取apk的名称
     */
    public String getApkName() {
        return apkName;
    }

    /**
     * 设置apk的名称
     */
    public DownloadManager setApkName(String apkName) {
        this.apkName = apkName;
        return this;
    }

    /**
     * 获取apk的保存路径
     */
    public String getDownloadPath() {
        return downloadPath;
    }

    /**
     * 设置apk的保存路径
     */
    public DownloadManager setDownloadPath(String downloadPath) {
        this.downloadPath = downloadPath;
        return this;
    }

    /**
     * 设置是否提示用户"当前已是最新版本"
     */
    public DownloadManager setShowNewerToast(boolean showNewerToast) {
        this.showNewerToast = showNewerToast;
        return this;
    }

    /**
     * 获取是否提示用户"当前已是最新版本"
     */
    public boolean isShowNewerToast() {
        return showNewerToast;
    }

    /**
     * 获取通知栏图片资源id
     */
    public int getSmallIcon() {
        return smallIcon;
    }

    /**
     * 设置通知栏图片资源id
     */
    public DownloadManager setSmallIcon(int smallIcon) {
        this.smallIcon = smallIcon;
        return this;
    }

    /**
     * 设置这个库的额外配置信息
     *
     * @see UpdateConfiguration
     */
    public DownloadManager setConfiguration(UpdateConfiguration configuration) {
        this.configuration = configuration;
        return this;
    }

    /**
     * 获取这个库的额外配置信息
     *
     * @see UpdateConfiguration
     */
    public UpdateConfiguration getConfiguration() {
        return configuration;
    }

    /**
     * 获取apk的versionName
     */
    public String getApkVersionName() {
        return apkVersionName;
    }

    /**
     * 设置apk的versionName
     */
    public DownloadManager setApkVersionName(String apkVersionName) {
        this.apkVersionName = apkVersionName;
        return this;
    }

    /**
     * 获取新版本描述信息
     */
    public String getApkDescription() {
        return apkDescription;
    }

    /**
     * 设置新版本描述信息
     */
    public DownloadManager setApkDescription(String apkDescription) {
        this.apkDescription = apkDescription;
        return this;
    }

    /**
     * 获取新版本文件大小
     */
    public String getApkSize() {
        return apkSize;
    }

    /**
     * 设置新版本文件大小
     */
    public DownloadManager setApkSize(String apkSize) {
        this.apkSize = apkSize;
        return this;
    }

    /**
     * 设置Android N Uri授权authorities
     */
    public String getAuthorities() {
        return authorities;
    }

    /**
     * 获取Android N Uri授权authorities
     */
    public DownloadManager setAuthorities(String authorities) {
        this.authorities = authorities;
        return this;
    }

    /**
     * 开始下载
     */
    public void download() {
        if (!checkParams()) {
            //参数设置出错....
            return;
        }
        if (checkVersionCode()) {
            //使用缓存目录不申请权限
            if (!downloadPath.equals(context.getExternalCacheDir().getPath())) {
                //检查权限
                if (!PermissionUtil.checkStoragePermission(context)) {
                    context.startActivity(new Intent(context, PermissionActivity.class));
                    return;
                }
            }
            context.startService(new Intent(context, DownloadService.class));
        } else {
            //对版本进行判断，是否显示升级对话框
            if (apkVersionCode > ApkUtil.getVersionCode(context)) {
                UpdateDialog dialog = new UpdateDialog(context);
                dialog.show();
            } else {
                if (showNewerToast) {
                    Toast.makeText(context, R.string.latest_version, Toast.LENGTH_SHORT).show();
                }
                LogUtil.e(TAG, "当前已是最新版本");
            }
        }
    }

    /**
     * 检查参数
     */
    private boolean checkParams() {
        if (TextUtils.isEmpty(apkUrl)) {
            LogUtil.e(TAG, "apkUrl can not be empty!");
            return false;
        }
        if (TextUtils.isEmpty(apkName)) {
            LogUtil.e(TAG, "apkName can not be empty!");
            return false;
        }
        if (!apkName.endsWith(Constant.APK_SUFFIX)) {
            LogUtil.e(TAG, "apkName must endsWith .apk!");
            return false;
        }
        /*
            这里需要注意，如果用户没有设置保存目录则使用缓存目录
            路径为:/storage/emulated/0/Android/data/ your packageName /cache
            如果使用的是缓存路径，则不申请内存权限
         */
        if (TextUtils.isEmpty(downloadPath)) {
            downloadPath = context.getExternalCacheDir().getPath();
        }
        if (smallIcon == -1) {
            LogUtil.e(TAG, "smallIcon can not be empty!");
            return false;
        }
        //如果用户没有进行配置，则使用默认的配置
        if (configuration == null) {
            configuration = new UpdateConfiguration();
        }
        return true;
    }

    /**
     * 检查设置的apkVersionCode 如果是大于1则使用内置的对话框
     * 如果小于等于1则直接启动服务下载
     */
    private boolean checkVersionCode() {
        //如果设置了小于的versionCode 你不是在写bug就是脑子瓦塌拉
        if (apkVersionCode < 1) {
            apkVersionCode = 1;
            LogUtil.e(TAG, "apkVersionCode can not be < 1 !");
            return true;
        }
        if (apkVersionCode > 1) {
            //设置了 VersionCode 则库中进行对话框逻辑处理
            if (TextUtils.isEmpty(apkDescription)) {
                LogUtil.e(TAG, "apkDescription can not be empty!");
            }
            return false;
        }
        //等于1的情况
        return true;
    }

    /**
     * 释放资源
     */
    public void release() {
        manager = null;
    }
}

File: appupdate/src/main/java/com/azhon/appupdate/manager/DownloadManager.kt
package com.azhon.appupdate.manager

import android.app.Activity
import android.app.Application
import android.app.NotificationChannel
import android.content.Intent
import android.util.Log
import android.widget.Toast
import androidx.annotation.LayoutRes
import androidx.fragment.app.FragmentActivity
import com.azhon.appupdate.R
import com.azhon.appupdate.base.BaseHttpDownloadManager
import com.azhon.appupdate.config.Constant
import com.azhon.appupdate.listener.LifecycleCallbacksAdapter
import com.azhon.appupdate.listener.OnButtonClickListener
import com.azhon.appupdate.listener.OnDownloadListener
import com.azhon.appupdate.service.DownloadService
import com.azhon.appupdate.util.ApkUtil
import com.azhon.appupdate.util.LogUtil
import com.azhon.appupdate.util.NotNullSingleVar
import com.azhon.appupdate.util.NullableSingleVar
import com.azhon.appupdate.view.PixelUpdateDialogFragment
import com.azhon.appupdate.view.SimpleUpdateDialog
import com.azhon.appupdate.view.UpdateDialogActivity
import com.azhon.appupdate.view.Win8UpdateDialogFragment
import java.io.Serializable

/**
 * ProjectName: AppUpdate
 * PackageName: com.azhon.appupdate.manager
 * FileName:    DownloadManager
 * CreateDate:  2022/4/7 on 10:36
 * Desc:
 *
 * @author azhon
 */
class DownloadManager private constructor(
    var config: DownloadConfig,
    internal var application: Application
) : Serializable {

    companion object {
        private const val TAG = "DownloadManager"

        @Volatile
        private var instance: DownloadManager? = null

        fun config(application: Application, block: DownloadConfig.() -> Unit): DownloadManager {
            val config = DownloadConfig(application)
            config.block()
            return getInstance(config, application)
        }

        fun isDownloading(): Boolean {
            return instance?.downloadState ?: false
        }

        fun existInstance(): DownloadManager? = instance

        internal fun getInstance(
            config: DownloadConfig? = null,
            application: Application? = null
        ): DownloadManager {
            if (instance != null && config != null) {
                instance!!.reConfig(config)
            }
            if (instance == null) {
                if (config == null || application == null) throw IllegalArgumentException("config or application is null")
                synchronized(this) {
                    instance ?: DownloadManager(config, application).also { instance = it }
                }
            }
            return instance!!
        }
    }

    private var apkVersionCode: Int = config.apkVersionCode

    var downloadState: Boolean = false

    /**
     * test whether can download
     */
    fun canDownload(): Boolean {
        val params = checkParams()
        val versionCodeCheck: Boolean =
            if (apkVersionCode == Int.MIN_VALUE) {
                true
            } else {
                apkVersionCode > ApkUtil.getVersionCode(application)
            }
        return (params && versionCodeCheck)

    }

    /**
     * download file without dialog
     */
    fun checkThenDownload() {
        if (canDownload()) {
            application.startService(Intent(application, DownloadService::class.java))
        } else {
            Log.e(TAG, "download: cannot download")
        }
    }

    /**
     * with out check whether can download, so you need to check it yourself
     */
    fun directDownload() {
        application.startService(Intent(application, DownloadService::class.java))
    }

    private fun checkParams(): Boolean {
        if (config.apkUrl.isEmpty()) {
            LogUtil.e(TAG, "apkUrl can not be empty!")
            return false
        }
        if (config.apkName.isEmpty()) {
            LogUtil.e(TAG, "apkName can not be empty!")
            return false
        }
        if (!config.apkName.endsWith(Constant.APK_SUFFIX)) {
            LogUtil.e(TAG, "apkName must endsWith .apk!")
            return false
        }
        if (config.smallIcon == -1) {
            LogUtil.e(TAG, "smallIcon can not be empty!");
            return false
        }
        if (config.apkDescription.isEmpty()) {
            LogUtil.e(TAG, "apkDescription can not be empty!")
            return false
        }
        Constant.AUTHORITIES = "${application.packageName}.fileProvider"
        return true
    }

    /**
     * when download not start,HttpManager maybe is null
     * in this case, will exec "then" block
     */
    fun cancel(then: () -> Unit = {}) {
        config.httpManager?.cancel() ?: then()
    }

    /**
     * release objects
     */
    internal fun release() {
        config.httpManager?.release()
        clearListener()
        instance = null
    }

    internal fun reConfig(config: DownloadConfig) {
        this.config.httpManager?.release()
        this.config = config
    }

    fun clearListener() {
        config.onButtonClickListener = null
        config.onDownloadListeners.clear()
    }

    class DownloadConfig constructor(application: Application) {
        /**
         * Apk download url
         */
        var apkUrl = ""

        /**
         * Apk file name on disk
         */
        var apkName = ""

        /**
         * The apk versionCode that needs to be downloaded
         * 如果不设置此值，将不检查当前应用的versioncode，直接下载安装更新
         */
        var apkVersionCode = Int.MIN_VALUE

        /**
         * The versionName of the dialog reality
         */
        var apkVersionName = ""

        /**
         * The file path where the Apk is saved
         * eg: /storage/emulated/0/Android/data/ your packageName /cache
         */
        var downloadPath = ApkUtil.getDefaultCachePath(application)

        /**
         * Notification icon resource
         */
        var smallIcon = -1

        /**
         * New version description information
         */
        var apkDescription = ""

        /**
         * Apk Size,Unit MB
         */
        var apkSize = ""

        /**
         * Apk md5 file verification(32-bit) verification repeated download
         */
        var apkMD5 = ""

        /**
         * Apk download manager
         */
        var httpManager: BaseHttpDownloadManager? by NullableSingleVar()

        /**
         * The following are unimportant filed
         */

        /**
         * adapter above Android O notification
         */
        var notificationChannel: NotificationChannel? = null

        /**
         * download listeners
         */
        val onDownloadListeners: MutableList<OnDownloadListener> by NotNullSingleVar(mutableListOf())

        /**
         * dialog button click listener
         */
        var onButtonClickListener: OnButtonClickListener? = null

        /**
         * Whether to show the progress of the notification
         */
        var showNotification = true

        /**
         * Whether the installation page will pop up automatically after the download is complete
         */
        var jumpInstallPage = true

        /**
         * Does the download start tip "Downloading a new version in the background..."
         */
        var showBgdToast = true

        /**
         * Whether to force an upgrade
         */
        var forcedUpgrade = false

        /**
         * Notification id
         */
        var notifyId = Constant.DEFAULT_NOTIFY_ID

        /**
         * View type
         */
        var viewType: Int = ViewType.Colorful


        fun enableLog(enable: Boolean): DownloadConfig {
            LogUtil.enable(enable)
            return this
        }

        fun registerDownloadListener(onDownloadListener: OnDownloadListener): DownloadConfig {
            this.onDownloadListeners.add(onDownloadListener)
            return this
        }

        fun registerButtonListener(onButtonClickListener: OnButtonClickListener): DownloadConfig {
            this.onButtonClickListener = onButtonClickListener
            return this
        }

        private fun getOneHttpManager(): BaseHttpDownloadManager {
            if (httpManager == null) {
                synchronized(this) {
                    if (httpManager == null) {
                        try {
                            httpManager = HttpDownloadManager(downloadPath)
                        } catch (e: Exception) {
                            Log.e(TAG, "getOneHttpManager: 重复赋值", e)
                        }
                    }
                }
            }
            return httpManager!!
        }

        /**
         * 真正的下载
         * @return Flow<DownloadStatus>
         */
        internal fun download() = getOneHttpManager().download(apkUrl, apkName)

        internal var viewConfig: DialogConfig = DialogConfig()

        /**
         * 配置视图
         */
        fun configDialog(block: DialogConfig.() -> Unit): DownloadConfig {
            viewConfig.block()
            return this
        }
    }

    class DialogConfig() {
        /**
         * custom layout id
         */
        var layoutId: Int = -1

        /**
         * whether to tip to user "Currently the latest version!"
         */
        var showNewerToast = false

        /**
         * dialog background Image resource
         */
        var dialogImage = -1

        /**
         * dialog button background color
         */
        var dialogButtonColor = -1

        /**
         * dialog button text color
         */
        var dialogButtonTextColor = -1

        /**
         * dialog progress bar color and progress-text color
         */
        var dialogProgressBarColor = -1
    }

    /**
     * 兼容旧方式
     * @property config DownloadConfig
     * @property dialogConfig DialogConfig
     * @constructor
     */
    class Builder constructor(activity: Activity) {
        val ctx = activity.application
        var config: DownloadManager.DownloadConfig =
            DownloadManager.DownloadConfig(application = ctx)
        var dialogConfig: DownloadManager.DialogConfig = DownloadManager.DialogConfig()

        fun apkUrl(apkUrl: String): Builder {
            config.apkUrl = apkUrl
            return this
        }

        fun downloadPath(path: String): Builder {
            config.downloadPath = path
            return this
        }

        fun apkName(apkName: String): Builder {
            config.apkName = apkName
            return this
        }

        fun apkVersionCode(apkVersionCode: Int): Builder {
            config.apkVersionCode = apkVersionCode
            return this
        }

        fun apkVersionName(apkVersionName: String): Builder {
            config.apkVersionName = apkVersionName
            return this
        }

        fun showNewerToast(showNewerToast: Boolean): Builder {
            dialogConfig.showNewerToast = showNewerToast
            return this
        }

        fun smallIcon(smallIcon: Int): Builder {
            config.smallIcon = smallIcon
            return this
        }

        fun apkDescription(apkDescription: String): Builder {
            config.apkDescription = apkDescription
            return this
        }

        fun apkSize(apkSize: String): Builder {
            config.apkSize = apkSize
            return this
        }

        fun apkMD5(apkMD5: String): Builder {
            config.apkMD5 = apkMD5
            return this
        }

        fun httpManager(httpManager: BaseHttpDownloadManager): Builder {
            config.httpManager = httpManager
            return this
        }

        fun notificationChannel(notificationChannel: NotificationChannel): Builder {
            config.notificationChannel = notificationChannel
            return this
        }

        fun onButtonClickListener(onButtonClickListener: OnButtonClickListener): Builder {
            config.onButtonClickListener = onButtonClickListener
            return this
        }

        fun onDownloadListener(onDownloadListener: OnDownloadListener): Builder {
            config.onDownloadListeners.add(onDownloadListener)
            return this
        }

        fun showNotification(showNotification: Boolean): Builder {
            config.showNotification = showNotification
            return this
        }

        fun jumpInstallPage(jumpInstallPage: Boolean): Builder {
            config.jumpInstallPage = jumpInstallPage
            return this
        }

        fun showBgdToast(showBgdToast: Boolean): Builder {
            config.showBgdToast = showBgdToast
            return this
        }

        fun forcedUpgrade(forcedUpgrade: Boolean): Builder {
            config.forcedUpgrade = forcedUpgrade
            return this
        }

        fun notifyId(notifyId: Int): Builder {
            config.notifyId = notifyId
            return this
        }

        fun dialogImage(dialogImage: Int): Builder {
            dialogConfig.dialogImage = dialogImage
            return this
        }

        fun dialogButtonColor(dialogButtonColor: Int): Builder {
            dialogConfig.dialogButtonColor = dialogButtonColor
            return this
        }

        fun dialogButtonTextColor(dialogButtonTextColor: Int): Builder {
            dialogConfig.dialogButtonTextColor = dialogButtonTextColor
            return this
        }

        fun dialogProgressBarColor(dialogProgressBarColor: Int): Builder {
            dialogConfig.dialogProgressBarColor = dialogProgressBarColor
            return this
        }

        fun dialogLayout(@LayoutRes layout: Int): Builder {
            dialogConfig.layoutId = layout
            return this
        }

        fun viewType(viewType: Int = ViewType.Colorful) {
            config.viewType = viewType
        }

        fun enableLog(enable: Boolean): Builder {
            config.enableLog(enable)
            return this
        }

        fun build(): DownloadManager {
            config.viewConfig = dialogConfig
            return DownloadManager.getInstance(config, application = ctx)
        }
    }

}

fun DownloadManager.download(activity: FragmentActivity) {
    if (canDownload()) {
        showUi(this, activity)
    } else {
        if (config.viewConfig.showNewerToast) {
            Toast.makeText(
                application, R.string.app_update_latest_version, Toast.LENGTH_SHORT
            ).show()
        }
    }
}

/**
 * 这个方法将会忽略配置中的viewType，使用默认UpdateDialogActivity进行下载
 * @receiver DownloadManager
 */
fun DownloadManager.download() {
    application.startActivity(
        Intent(application, UpdateDialogActivity::class.java)
            .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
    )
    // Fix memory leak
    application.registerActivityLifecycleCallbacks(object :
        LifecycleCallbacksAdapter() {
        override fun onActivityDestroyed(activity: Activity) {
            super.onActivityDestroyed(activity)
            if (this.javaClass.name == activity.javaClass.name) {
                clearListener()
            }
        }
    })

    if (!canDownload()) {
        if (config.viewConfig.showNewerToast) {
            Toast.makeText(
                application, R.string.app_update_latest_version, Toast.LENGTH_SHORT
            ).show()
        }
    }
}
